### ISSUE TITLE
primarykey(_:) does not allow specifying a schema name

### ISSUE THREAD

**Issue by barnettben** [groue/GRDB.swift#1463]
When using the table introspection function `primaryKey(_:)`, you can't specify a schema name, so it returns details for the first matching table name. This can be a problem with multiple attached databases with conflicting table names.

Other introspection functions (`indexes(on:)`, `foreignKeys(on:)` etc) also show this behaviour.

I am happy to submit a pull request for this, but would like to check if this is something that merits changing and if so what the change should look like.

Options I have considered are:
1. Adding an optional second parameter to `primaryKey(_:)`:
    `public func primaryKey(_ tableName: String, in schemaName: String? = nil) throws -> PrimaryKeyInfo`
    or
    `public func primaryKey(_ tableName: String, in schemaName: String = "main") throws -> PrimaryKeyInfo`
2. Adding one of the above with a required second parameter as a separate function in addition to the current one.

But I'm not sure which one - or if something else - would be preferred.

---

### What did you do?

Take two database files:
- `schema_v1.db` has the table `CREATE TABLE example (id INTEGER PRIMARY KEY, name TEXT);`
- `schema_v2.db` has the table `CREATE TABLE example (id INTEGER, siteID INTEGER, name TEXT, PRIMARY KEY(id, siteID));`

Create a database queue with `schema_v1.db` and attached the second file:
- `ATTACH DATABASE /path/to/schema_v2.db AS v2;`

Finally, query the primary keys for these tables:

```swift
try dbQueue.read { db in
  try db.primaryKey("example").columns
}
// ["id"]
```

But because I cannot specify a schema name for `primaryKey(_:)`, I am unable to find the primary keys for `v2.example`.

### What did you expect to happen?

To be able to specify a schema name when fetching primary key information.

### What happened instead?

I received information for the first available table matching that name

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 6.23.0
**Installation method:** SPM
**Xcode version:** 15.0
**Swift version:** 5.9
**Platform(s) running GRDB:** macOS
**macOS version running Xcode:** 14.0

### Demo Project

A quick demo project which shows only being able to fetch primary key details for the first table matching a given name:
[PrimaryKeyDemo.zip](https://github.com/groue/GRDB.swift/files/13538729/PrimaryKeyDemo.zip)

**Comment by groue** [comment]
Hello @barnettben, happy to see you contribute again :-)

Thanks for opening this great issue. There already exists a support for database schemas in internal apis, but they have never surfaced the public api yet.

All schema introspection methods, such as [`primaryKey(_:)`](https://github.com/groue/GRDB.swift/blob/v6.23.0/GRDB/Core/Database%2BSchema.swift#L278-L285), iterate known schemas in the [SQLite resolution order](https://www.sqlite.org/lang_naming.html) in order to find a matching table. With those apis, everything works as expected as long as the main database, the temp database, and attached database, have distinct object names.

> I am happy to submit a pull request for this

It would be welcome.

> [...] but would like to check if this is something that merits changing and if so what the change should look like.

Yes, you're very right :-)

> 1. Adding an optional second parameter to `primaryKey(_:)`:
>    `public func primaryKey(_ tableName: String, in schemaName: String? = nil) throws -> PrimaryKeyInfo`
>    or
>    `public func primaryKey(_ tableName: String, in schemaName: String = "main") throws -> PrimaryKeyInfo`
>
> 2. Adding one of the above with a required second parameter as a separate function in addition to the current one.

The `primaryKey(_ tableName: String, in schemaName: String = "main")` version would be a breaking change. With a "main" default schema, it would fail finding the primary key for a matching table in the temp and attached databases, as it currently does.

But I do wonder what's the best way to surface schemas in the public api. Isn't it the regular identified table names, just as in the SQLite syntax (`main.player`, `temp.player`, , `v2.player`)?

```swift
let pk = try db.primaryKey("v2.player")
```

If GRDB would start dealing with identified table names (and views, etc), we would gain a lot of features without any API change.

Record types would just provide a `databaseTableName` static property, just as they do today. No need for an extra `databaseSchemaName` property:

```swift
struct PlayerV2: TableRecord {
    static let databaseTableName = "v2.player" // NEW!
}

// SELECT COUNT(*) FROM v2.player;
let playerCount = try PlayerV2.fetchCount(db)
```

The `Table` type that feeds the query builder would not need an extra `schemaName` argument:

```swift
let table = Table("v2.player") // NEW!

// SELECT COUNT(*) FROM v2.player;
let playerV2Count = try table.fetchCount(db)
```

Schema introspection methods would not need to change their signature:

```swift
let primaryKey = try db.primaryKey("v2.player") // NEW!
let indexes = try db.indexes(on: "v2.player") // NEW!
```

In my current opinion, that would be ideal. A feature well integrated throughout the library, without any surprising inconsistency between GRDB components. A public interface that directly fits the expectations of developers who happen to know about SQLite schemas. A public interface that is not difficult to learn for other people.

This does not come without difficulty, unfortunately. Some users may currently use table names that look like identified names, and GRDB interprets them as tables in the main schema 😖:

```swift
struct TempPlayer: TableRecord {
    static let databaseTableName = "temp.player"
}

// The current behavior
// SELECT COUNT(*) FROM "temp.player"; -- The 'temp.player' table from the main schema
let count = try TempPlayer.fetchCount(db)
```

But what are the odds? Is there any way we could we help those users if we'd introduce the breaking change?

---

As a conclusion, before we consider adding a `schemaName` argument to schema introspection methods, it would be good to make a clear opinion for or against adding direct support for identified names. If it's a clear no, then I suppose the extra `schemaName` argument would be the way to go.

It would be a great contribution to help settle this case!

**Comment by barnettben** [comment]
Thank you for the positive response!

I agree that your proposed method appears friendlier to (most) existing users and would require fewer API changes.

My thoughts on handling tables with `.` in them:

~~People doing that are doing it wrong~~ :grin:

- Quotes?
To account for any users already using entity names with dots in, you could allow parts of names to be quoted, as SQLite already does:
```swift
struct TempPlayer: TableRecord {
    static let databaseTableName = "main.\"temp.player\""
}
```
But then this would likely become a problem internally as previously-simple strings would need parsing into different parts. And it looks weird. I don't like it.

- Splitting?
Like you've said above, You could instead split only on the first `.`, like so:
```swift
"temp.player" // Schema: "temp", table: "player"
"temp.temp.player" // Schema: "temp", table: "temp.player"
"main.temp.player" // Schema: "main", table: "temp.player"
```
It's still a breaking change for some users, but at least a simple fix.
It also prevents you from having a `.` in the schema name, which is technically allowed.

---

The only way I can think to help people using table names with dots in currently is to print a message to the console when a table name with dots in is used.
You could remove the message after a few versions to give time to adapt. Alternatively you could print a message warning of an upcoming change, then introduce the changes we are discussing after a period of time.
This is only a short-term help though. On the other hand, is that naming convention a common practice and will it affect many people? I don't know!

---

I think it depends on how you want to balance supporting allowable SQL syntax vs ease of development.
It looks like the only way to 100% support what SQLite supports is separate parameters or string parsing. But on the other hand that's a lot of changes for perhaps a small subset of users.
For my own needs, the splitting option would be an acceptable trade-off.
I'm sorry, but even though I suggested this change, I don't think I can make that choice for everybody!

**Comment by groue** [comment]
Thank you @barnettben for opening my eyes :-) Supporting dotted table names is not as desirable a feature as I initially thought. Your initial idea is the best! 👍 

Parsing is indeed the key, and it is weird, I agree. Now I understand that it will be complicated, difficult to maintain, and even baffle user expectations at some point. Take the user-provided string `"temp.group"`. It is not a valid SQLite identified table name. The valid SQLite identifier is `"temp.\"group\""`. Dealing with such a case can not be solved in a simple way... Forcing the user to provide the quoted string is not user-friendly. We currently support the plain `"group"` table name, without quoting. In the end, we'd need a lot of cheesy magic in order to support `"temp.group"`.

---

You suggestion of adding a default `schemaName: String? = nil` argument looks quite fine to me:

```diff
-func primaryKey(_ tableName: String) throws -> PrimaryKeyInfo
+func primaryKey(_ tableName: String, in schemaName: String? = nil) throws -> PrimaryKeyInfo
```

Let's focus of schema introspection methods for a start, and leave other components aside (the query builder with `Table` and `TableRecord`, database observation, external-content full-text tables...).

It looks like the following methods could enjoy the extra `schemaName` argument:

- `func columns(in: String) throws -> [ColumnInfo]`
- `func foreignKeys(on: String) throws -> [ForeignKeyInfo]`
- `func indexes(on: String) throws -> [IndexInfo]`
- `func primaryKey(String) throws -> PrimaryKeyInfo`
- `func tableExists(String) throws -> Bool`
- `func triggerExists(String) throws -> Bool`
- `func viewExists(String) throws -> Bool`

This one below will only be necessary when `Table` and `TableRecord` support explicit schemas, so we can leave it aside:

- `func table<Columns>(String, hasUniqueKey: Columns) throws -> Bool`

The two methods that deal with foreign key violations are intended to be used from migrations. I don't use attached tables enough to know if they should be included or not in our set of extended methods. I leave this decision to your good taste :-)

- `func foreignKeyViolations(in: String) throws -> RecordCursor<ForeignKeyViolation>`
- `func checkForeignKeys(in: String) throws`

Does this plan sound good to you? I'll happily accept a PR, and I will help if you have any question!

Here is an advice that is not (yet) written in CONTRIBUTING.md. Since some method signatures change, we need to update the DocC documentation. To build it, open GRDB as an SPM package (open `Package.swift`, not `GRDB.xcworkspace`). The DocC compiler will emit warnings for references that need to be updated. I actually think that all changes related to this issue can be fully worked from the SPM package.

The tests of the four first methods are located in `ColumnInfoTests.swift`, `ForeignKeyInfoTests.swift`, `IndexInfoTests.swift`, `PrimaryKeyInfoTests.swift`.

The methods `tableExists`, `triggerExists`, `viewExists` are not currently tested on their own. They are indirectly tested in the tests of higher level entities. I recommend extending `DatabaseQueueSchemaCacheTests`. It is also in this file that you'll find sample code for testing attached databases.

**Comment by barnettben** [comment]
Thank you @groue, that sounds good. I will work on this and submit a PR.

I'm also not too familiar with the foreign key violation checks, but will read up on them and appreciate your offer of help if I have questions.

**Comment by groue** [comment]
Fixed by #1466
